File: tools/tour/evidence.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'base64'
require 'digest'
require 'fileutils'
require 'json'
require 'open3'
require 'rbconfig'
require 'tmpdir'
require 'tempfile'

module TourEvidence
  SCHEMA = 'tour-evidence/v2'
  MODES = %w[baseline interpreted compiled].freeze
  STAGES = %w[transpile build run].freeze

  module_function

  def canonical(object)
    JSON.generate(deep_sort(object))
  end

  def deep_sort(object)
    case object
    when Hash then object.keys.sort.to_h { |key| [key, deep_sort(object[key])] }
    when Array then object.map { |value| deep_sort(value) }
    else object
    end
  end

  def sha(bytes)
    Digest::SHA256.hexdigest(bytes)
  end

  def normalize(bytes, normalizer)
    out, err, status = Open3.capture3(RbConfig.ruby, normalizer, stdin_data: bytes, binmode: true)
    raise "normalizer wrote stderr: #{err}" unless err.empty?
    return { 'valid_utf8' => false, 'bytes' => nil, 'sha256' => nil, 'base64' => nil } unless status.success?
    { 'valid_utf8' => true, 'bytes' => out.bytesize, 'sha256' => sha(out), 'base64' => Base64.strict_encode64(out) }
  end

  # Every command gets a new process group. On deadline, and again after wait,
  # the entire group is swept so descendants cannot retain capture descriptors.
  def capture(argv, chdir:, timeout:, env: {})
    out = Tempfile.new('tour-out'); err = Tempfile.new('tour-err')
    out.binmode; err.binmode
    result = { 'spawned' => false, 'state' => 'launch_failure', 'exit' => nil }
    begin
      pid = Process.spawn(env, *argv, chdir: chdir, in: File::NULL, out: out, err: err, pgroup: true)
      result['spawned'] = true
      deadline = Process.clock_gettime(Process::CLOCK_MONOTONIC) + timeout
      status = nil
      until status
        waited, status = Process.waitpid2(pid, Process::WNOHANG)
        break if waited
        if Process.clock_gettime(Process::CLOCK_MONOTONIC) >= deadline
          result['state'] = 'deadline'
          begin Process.kill('TERM', -pid); rescue Errno::ESRCH; end
          sleep 0.1
          begin Process.kill('KILL', -pid); rescue Errno::ESRCH; end
          _, status = Process.waitpid2(pid)
          break
        end
        sleep 0.01
      end
      if result['state'] != 'deadline'
        result['state'] = 'exited'
        result['exit'] = status.exitstatus || 128 + status.termsig
      end
      begin Process.kill('KILL', -pid); rescue Errno::ESRCH; end
    rescue SystemCallError => e
      err.write("#{e.class}: #{e.message}\n")
    ensure
      out.flush; err.flush; out.rewind; err.rewind
      result['stdout'] = out.read.b
      result['stderr'] = err.read.b
      out.close!; err.close!
    end
    result
  end

  def derived(raw, normalizer)
    {
      'stdout' => normalize(raw.fetch('stdout'), normalizer),
      'stderr' => normalize(raw.fetch('stderr'), normalizer)
    }
  end

  def agent_banner?(raw)
    raw.fetch('stderr').match?(/^bashy: .* detected, and this repo has no agent config /)
  end

  def outcome(attempt, baseline = nil)
    return 'FAIL:launch_failure' unless attempt['spawned']
    return 'FAIL:deadline' if attempt['state'] == 'deadline'
    return "FAIL:exit:#{attempt['exit']}" unless attempt['exit'] == 0
    return 'FAIL:invalid_utf8' unless attempt.dig('normalized', 'stdout', 'valid_utf8') && attempt.dig('normalized', 'stderr', 'valid_utf8')
    if baseline && %w[stdout stderr].any? { |stream| attempt.dig('normalized', stream, 'sha256') != baseline.dig('normalized', stream, 'sha256') }
      return 'FAIL:mismatch'
    end
    'PASS'
  end

  def root(records)
    sha(records.map { |record| canonical(record) }.join("\n") + "\n")
  end

  # Command construction for the two single-shot modes. Shared by the runner
  # (which executes it) and the validator (which replays it independently) so
  # the two can never drift into checking a different command than the one
  # that actually ran.
  def command_for(mode, item, bashy:, go_bin:, local:)
    case mode
    when 'baseline'
      [go_bin, item['applicability'] == 'build_only_go_program' ? 'build' : 'run', local]
    when 'interpreted'
      [bashy, '--bashpp', *(item['applicability'] == 'build_only_go_program' ? ['-n'] : []), local]
    else
      raise ArgumentError, "command_for does not handle mode #{mode}"
    end
  end

  # Compiled mode is a three-stage pipeline: transpile the pinned Go source
  # with bashy, build the transpiled output with the exact pinned Go 1.27
  # toolchain, then execute the resulting binary. The recorded raw/state/exit
  # belong to whichever stage is authoritative: the first stage that fails to
  # produce its declared artifact, or — when every stage succeeds — the
  # executed binary itself. A "compiled" observation is therefore never just
  # the transpiler's own output; it is the real behavior of the program bashy
  # produced, exactly like the differential schema (`bpp_compiled:
  # transpile-build-run`) declares.
  def compiled_pipeline(bashy:, go_bin:, source:, module_dir:, workdir:, timeout:, env: {})
    base = File.basename(source, '.go')
    transpiled = File.join(workdir, "#{base}.transpiled.go")
    binary = File.join(workdir, "#{base}.bin")
    FileUtils.rm_f([transpiled, binary])
    command = {
      'transpile' => [bashy, 'transpile', source, '-o', transpiled],
      'build' => [go_bin, 'build', '-o', binary, transpiled],
      'run' => [binary]
    }

    transpile_raw = capture(command['transpile'], chdir: module_dir, timeout: timeout, env: env)
    unless transpile_raw['spawned'] && transpile_raw['state'] == 'exited' && transpile_raw['exit'] == 0 && File.file?(transpiled)
      return { 'command' => command, 'stage' => 'transpile', 'raw' => transpile_raw }
    end

    build_raw = capture(command['build'], chdir: module_dir, timeout: timeout, env: env.merge('GOTOOLCHAIN' => 'local'))
    unless build_raw['spawned'] && build_raw['state'] == 'exited' && build_raw['exit'] == 0 && File.executable?(binary)
      return { 'command' => command, 'stage' => 'build', 'raw' => build_raw }
    end

    run_raw = capture(command['run'], chdir: module_dir, timeout: timeout, env: {})
    { 'command' => command, 'stage' => 'run', 'raw' => run_raw }
  end

  # Materializes the same scratch module (go.mod/go.sum for the pinned
  # helper, every pinned tour source copied and byte/sha verified) that a
  # run needs. Used both to produce evidence and to replay it, so a replay
  # executes against an identically-constructed tree.
  def materialize_module(mod_dir, tour_root:, inventory:, go_version:, helper:)
    FileUtils.mkdir_p(mod_dir)
    File.write(File.join(mod_dir, 'go.mod'), "module tour.evidence.local\n\ngo #{go_version.delete_prefix('go')}\n\nrequire #{helper[0]} #{helper[1]}\n")
    File.write(File.join(mod_dir, 'go.sum'), "#{helper[0]} #{helper[1]} #{helper[4]}\n#{helper[0]} #{helper[1]}/go.mod #{helper[3]}\n")
    # Published Bashy 0.20 predates BASHY_HINTS suppression for its one-time
    # startup advertisement. A deterministic agent-config marker prevents that
    # provenance-neutral, driver-specific line from entering evidence. It is
    # not Go source and does not alter any program under test.
    File.write(File.join(mod_dir, 'AGENTS.md'), "# Hermetic Tour evidence workspace\n")
    File.chmod(0o444, File.join(mod_dir, 'AGENTS.md'))
    license_source = File.join(tour_root, 'LICENSE')
    raise "missing upstream LICENSE #{license_source}" unless File.file?(license_source)
    license_bytes = File.binread(license_source)
    license_local = File.join(mod_dir, 'LICENSE')
    File.binwrite(license_local, license_bytes)
    File.chmod(0o444, license_local)
    inventory.each do |item|
      source = File.join(tour_root, item['path'])
      raise "missing source #{source}" unless File.file?(source)
      raise "source mode is not read-only 444: #{item['path']}" unless File.stat(source).mode & 0o777 == 0o444
      bytes = File.binread(source)
      raise "source pin mismatch #{item['path']}" unless bytes.bytesize == item['bytes'] && sha(bytes) == item['sha256']
      local = File.join(mod_dir, item['path'])
      FileUtils.mkdir_p(File.dirname(local))
      File.binwrite(local, bytes)
      File.chmod(0o444, local)
    end
  end

  # Parses `bashy --version`-style output (e.g.
  # "bashy, GNU Bash 5.3 compatible, version 5.3.0(1)-bashy-0.20.0" for a
  # published release, or "...-bashy-dev (da8deb9-dirty)" for a workspace
  # build) into a provenance identity that can be independently
  # re-derived — never trusted as free-form text. A dev build (published:
  # false) or a dirty tree (dirty: true) cannot satisfy the "clean published
  # reproducible" requirement, whatever the manifest claims about them.
  def bashy_identity(version_line)
    m = version_line.to_s.match(/-bashy-(dev|\d[\w.]*)(?:\s+\(([0-9a-f]{7,40})(-dirty)?\))?\z/)
    return { 'revision' => nil, 'published' => false, 'dirty' => true } unless m
    tag, commit, dirty_suffix = m[1], m[2], m[3]
    if tag == 'dev'
      { 'revision' => commit, 'published' => false, 'dirty' => true }
    else
      { 'revision' => tag, 'published' => true, 'dirty' => !dirty_suffix.nil? }
    end
  end
end

File: tools/tour/evidence.go
// Sprint 155 / Story S155.9 / Story-ID 43af37063b09.
//
// The superseded tour-evidence/v2 runner library — the port of
// tools/tour/evidence.rb (Sprint 98 / Story #4). Its ledger,
// tests/tour/evidence.jsonl, is retained unchanged as historical failure
// evidence and is still validated by `tour validate-evidence` (harness-wired).
package main

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
	"syscall"
	"time"
)

const (
	evidenceSchema = "tour-evidence/v2"
)

var (
	evidenceStages   = []string{"transpile", "build", "run"}
	agentBannerRE    = regexp.MustCompile(`(?m)^bashy: .* detected, and this repo has no agent config `)
	bashyIdentityRE  = regexp.MustCompile(`-bashy-(dev|\d[\w.]*)(?:\s+\(([0-9a-f]{7,40})(-dirty)?\))?\z`)
	unsafePathCharRE = regexp.MustCompile(`[^\w./]`)
)

// evidenceNormalize returns the tour-evidence normalized record, which also
// carries the normalized bytes in base64.
func evidenceNormalize(raw []byte) map[string]any {
	out, ok := normalizeV1(raw)
	if !ok {
		return map[string]any{"valid_utf8": false, "bytes": nil, "sha256": nil, "base64": nil}
	}
	return map[string]any{"valid_utf8": true, "bytes": int64(len(out)), "sha256": sha256hex(out), "base64": b64(out)}
}

// EvidenceCapture is TourEvidence.capture's result.
type EvidenceCapture struct {
	Spawned        bool
	State          string
	Exit           any
	Stdout, Stderr []byte
}

// evidenceCapture: every command gets a new process group. On deadline, and
// again after wait, the entire group is swept so descendants cannot retain
// capture descriptors.
func evidenceCapture(argv []string, chdir string, timeout float64, env map[string]string) EvidenceCapture {
	result := EvidenceCapture{Spawned: false, State: "launch_failure", Exit: nil}
	out, err := os.CreateTemp("", "tour-out")
	if err != nil {
		return result
	}
	defer os.Remove(out.Name())
	defer out.Close()
	errf, err := os.CreateTemp("", "tour-err")
	if err != nil {
		return result
	}
	defer os.Remove(errf.Name())
	defer errf.Close()
	devnull, _ := os.Open(os.DevNull)
	defer devnull.Close()

	cmd := exec.Command(argv[0], argv[1:]...)
	cmd.Path = argv[0]
	cmd.Dir = chdir
	// Like Process.spawn(env, ...) without unsetenv_others: `env` overlays the
	// inherited environment.
	cmd.Env = overlayEnv(env)
	cmd.Stdin = devnull
	cmd.Stdout = out
	cmd.Stderr = errf
	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
	if err := cmd.Start(); err != nil {
		fmt.Fprintf(errf, "Errno::ENOENT: %s\n", err.Error())
	} else {
		result.Spawned = true
		pid := cmd.Process.Pid
		deadline := time.Now().Add(time.Duration(timeout * float64(time.Second)))
		done := make(chan struct{})
		go func() {
			cmd.Wait()
			close(done)
		}()
		waited := false
		for !waited {
			select {
			case <-done:
				waited = true
			default:
				if time.Now().After(deadline) {
					result.State = "deadline"
					syscall.Kill(-pid, syscall.SIGTERM)
					time.Sleep(100 * time.Millisecond)
					syscall.Kill(-pid, syscall.SIGKILL)
					<-done
					waited = true
				} else {
					time.Sleep(10 * time.Millisecond)
				}
			}
		}
		if result.State != "deadline" {
			result.State = "exited"
			if ws, ok := cmd.ProcessState.Sys().(syscall.WaitStatus); ok {
				if ws.Exited() {
					result.Exit = int64(ws.ExitStatus())
				} else if ws.Signaled() {
					result.Exit = int64(128 + int(ws.Signal()))
				}
			}
		}
		syscall.Kill(-pid, syscall.SIGKILL)
	}
	result.Stdout, _ = os.ReadFile(out.Name())
	result.Stderr, _ = os.ReadFile(errf.Name())
	return result
}

func evidenceDerived(stdout, stderr []byte) map[string]any {
	return map[string]any{"stdout": evidenceNormalize(stdout), "stderr": evidenceNormalize(stderr)}
}

func agentBanner(stderr []byte) bool {
	return agentBannerRE.Match(stderr)
}

// evidenceOutcome mirrors TourEvidence.outcome.
func evidenceOutcome(attempt map[string]any, baseline map[string]any) string {
	if !truthy(attempt["spawned"]) {
		return "FAIL:launch_failure"
	}
	if attempt["state"] == "deadline" {
		return "FAIL:deadline"
	}
	if !jsonEqual(attempt["exit"], int64(0)) {
		return "FAIL:exit:" + toS(attempt["exit"])
	}
	if !truthy(dig(attempt, "normalized", "stdout", "valid_utf8")) || !truthy(dig(attempt, "normalized", "stderr", "valid_utf8")) {
		return "FAIL:invalid_utf8"
	}
	if baseline != nil {
		for _, stream := range []string{"stdout", "stderr"} {
			if !jsonEqual(dig(attempt, "normalized", stream, "sha256"), dig(baseline, "normalized", stream, "sha256")) {
				return "FAIL:mismatch"
			}
		}
	}
	return "PASS"
}

// evidenceCommandFor: command construction for the two single-shot modes.
func evidenceCommandFor(mode string, applicability string, bashy, goBin, local string) []string {
	switch mode {
	case "baseline":
		verb := "run"
		if applicability == "build_only_go_program" {
			verb = "build"
		}
		return []string{goBin, verb, local}
	case "interpreted":
		argv := []string{bashy, "--bashpp"}
		if applicability == "build_only_go_program" {
			argv = append(argv, "-n")
		}
		return append(argv, local)
	}
	panic("command_for does not handle mode " + mode)
}

// EvidencePipeline is the compiled pipeline's outcome: the authoritative stage
// and its raw capture.
type EvidencePipeline struct {
	Command map[string]any
	Stage   string
	Raw     EvidenceCapture
}

// evidenceCompiledPipeline: transpile with bashy, build with the exact pinned
// Go, execute the resulting binary. The recorded raw/state/exit belong to
// whichever stage is authoritative.
func evidenceCompiledPipeline(bashy, goBin, source, moduleDir, workdir string, timeout float64, env map[string]string) EvidencePipeline {
	base := strings.TrimSuffix(filepath.Base(source), ".go")
	transpiled := filepath.Join(workdir, base+".transpiled.go")
	binary := filepath.Join(workdir, base+".bin")
	os.Remove(transpiled)
	os.Remove(binary)
	command := map[string]any{
		"transpile": anyList([]string{bashy, "transpile", source, "-o", transpiled}),
		"build":     anyList([]string{goBin, "build", "-o", binary, transpiled}),
		"run":       anyList([]string{binary}),
	}
	transpileRaw := evidenceCapture(strList(command["transpile"]), moduleDir, timeout, env)
	if !(transpileRaw.Spawned && transpileRaw.State == "exited" && jsonEqual(transpileRaw.Exit, int64(0)) && fileExists(transpiled)) {
		return EvidencePipeline{command, "transpile", transpileRaw}
	}
	buildEnv := map[string]string{}
	for k, v := range env {
		buildEnv[k] = v
	}
	buildEnv["GOTOOLCHAIN"] = "local"
	buildRaw := evidenceCapture(strList(command["build"]), moduleDir, timeout, buildEnv)
	if !(buildRaw.Spawned && buildRaw.State == "exited" && jsonEqual(buildRaw.Exit, int64(0)) && isExecutable(binary)) {
		return EvidencePipeline{command, "build", buildRaw}
	}
	runRaw := evidenceCapture(strList(command["run"]), moduleDir, timeout, map[string]string{})
	return EvidencePipeline{command, "run", runRaw}
}

// evidenceMaterializeModule builds the scratch module (go.mod/go.sum for the
// pinned helper, every pinned tour source copied and byte/sha verified) that
// both the runner and the replay validator use.
func evidenceMaterializeModule(modDir, tourRoot string, inventory []Item, goVersion string, helper []string) error {
	if err := os.MkdirAll(modDir, 0o755); err != nil {
		return err
	}
	if err := os.WriteFile(filepath.Join(modDir, "go.mod"), []byte(fmt.Sprintf("module tour.evidence.local\n\ngo %s\n\nrequire %s %s\n", strings.TrimPrefix(goVersion, "go"), helper[0], helper[1])), 0o644); err != nil {
		return err
	}
	if err := os.WriteFile(filepath.Join(modDir, "go.sum"), []byte(fmt.Sprintf("%s %s %s\n%s %s/go.mod %s\n", helper[0], helper[1], helper[4], helper[0], helper[1], helper[3])), 0o644); err != nil {
		return err
	}
	if err := writeReadOnly(filepath.Join(modDir, "AGENTS.md"), []byte("# Hermetic Tour evidence workspace\n")); err != nil {
		return err
	}
	licenseSource := filepath.Join(tourRoot, "LICENSE")
	if !fileExists(licenseSource) {
		return fmt.Errorf("missing upstream LICENSE %s", licenseSource)
	}
	if err := writeReadOnly(filepath.Join(modDir, "LICENSE"), readFile(licenseSource)); err != nil {
		return err
	}
	for _, item := range inventory {
		source := filepath.Join(tourRoot, item.Path)
		st, err := os.Stat(source)
		if err != nil || !st.Mode().IsRegular() {
			return fmt.Errorf("missing source %s", source)
		}
		if st.Mode().Perm() != 0o444 {
			return fmt.Errorf("source mode is not read-only 444: %s", item.Path)
		}
		bytes := readFile(source)
		if int64(len(bytes)) != item.Bytes || sha256hex(bytes) != item.SHA256 {
			return fmt.Errorf("source pin mismatch %s", item.Path)
		}
		local := filepath.Join(modDir, item.Path)
		if err := os.MkdirAll(filepath.Dir(local), 0o755); err != nil {
			return err
		}
		if err := writeReadOnly(local, bytes); err != nil {
			return err
		}
	}
	return nil
}

// bashyIdentity parses `bashy --version` output into a provenance identity.
func bashyIdentity(versionLine string) map[string]any {
	m := bashyIdentityRE.FindStringSubmatch(versionLine)
	if m == nil {
		return map[string]any{"revision": nil, "published": false, "dirty": true}
	}
	tag, commit, dirty := m[1], m[2], m[3]
	if tag == "dev" {
		var rev any
		if commit != "" {
			rev = commit
		}
		return map[string]any{"revision": rev, "published": false, "dirty": true}
	}
	return map[string]any{"revision": tag, "published": true, "dirty": dirty != ""}
}

// evidenceInventory: the executable rows exactly as the evidence tools read them.
func evidenceInventory(path string) ([]Item, string) {
	lines := []string{}
	items := []Item{}
	for _, line := range strings.Split(strings.TrimSuffix(string(readFile(path)), "\n"), "\n") {
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		lines = append(lines, line)
		f := strings.Split(line, "\t")
		if !containsString(executableApplicabilities, field(f, 3)) {
			continue
		}
		items = append(items, Item{Path: field(f, 0), Applicability: field(f, 3), Bytes: mustInt(field(f, 6)), SHA256: field(f, 7)})
	}
	return items, sha256hex([]byte(strings.Join(lines, "\n") + "\n"))
}

func firstDataRow(path string) []string {
	for _, row := range tsvRowsLoose(path) {
		return row
	}
	return nil
}

func hostGoosGoarch() (string, string) {
	goos := strings.ToLower(shellOutput(nil, "uname", "-s"))
	goarch := shellOutput(nil, "uname", "-m")
	return goos, goarch
}

func overlayEnv(env map[string]string) []string {
	merged := map[string]string{}
	for _, kv := range os.Environ() {
		k, v, _ := strings.Cut(kv, "=")
		merged[k] = v
	}
	for k, v := range env {
		merged[k] = v
	}
	return envSlice(merged)
}
